src/main/java/servlets/ReceiptAnalysis.java

```diff
@@ -189,14 +205,89 @@ private static AnalysisResults analyzeImage(ByteString imageBytes)
     return categories;
   }
 
-  /*
+  /**
    * Parse category strings into more natural categories
    * e.g. "/Food & Drink/Restaurants" becomes "Food & Drink" and "Restaurants"
    */
   private static Stream<String> parseCategory(ClassificationCategory category) {
     return Stream.of(category.getName().substring(1).split("/"));
   }
 
+  /**
+   * Checks the provided tokens for a date that can be parsed. If one is found, it is added to the
+   * builder as a timestamp.
+   */
+  private static void checkForParsableDate(
+      AnalysisResults.Builder analysisBuilder, Stream<String> tokens) {
+    Stream<String> dates = tokens.filter(ReceiptAnalysis::isDate);
+    // Assume that the first date on the receipt is the transaction date.
+    Optional<String> firstDate = dates.findFirst();
+
+    firstDate.ifPresent(date -> ReceiptAnalysis.addDateIfValid(analysisBuilder, date));
+  }
+
+  /**
+   * Adds a valid date to the builder as a timestamp. If the date has an invalid month or day, then
+   * nothing is added.
+   */
+  private static void addDateIfValid(AnalysisResults.Builder analysisBuilder, String date) {
+    String separator = date.contains("-") ? "-" : "/";
+    String formatterPattern = "M" + separator + "d" + separator;
+
+    // Determine if the date has 2 or 4 digits for the year
+    if (date.lastIndexOf(separator) + 3 == date.length()) {
+      formatterPattern += "yy";
+    } else {
+      formatterPattern += "yyyy";
+    }
+
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern(formatterPattern);
+
+    try {
+      ZonedDateTime dateAndTime = LocalDate.parse(date, formatter).atStartOfDay(ZoneOffset.UTC);
+      dateAndTime = fixYearIfInFuture(dateAndTime);
+
+      long timestamp = dateAndTime.toInstant().toEpochMilli();
+      analysisBuilder.setTimestamp(timestamp);
+    } catch (DateTimeParseException e) {
+      // Invalid month or day
+      return;
+    }
+  }
+
+  /**
+   * DateTimeFormatter assumes that two-digit years are in the 2000s. This method checks if the
+   * given date is after the current date and if so, subtracts 100 years to move it to the 1900s.
+   */
+  private static ZonedDateTime fixYearIfInFuture(ZonedDateTime dateAndTime) {
```

Comment:
hopefully no one tries to upload a receipt that old either way lol! but good edge case